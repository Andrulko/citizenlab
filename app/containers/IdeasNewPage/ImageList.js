/**
*
* AttachmentList
*
*/

import React, { PropTypes } from 'react';
import { Label, Row, Column } from 'components/Foundation';
import styled from 'styled-components';
import { FormattedMessage } from 'react-intl';
import Image from './Image';
import messages from './messages';

class ImageList extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  constructor() {
    super();

    this.state = {
      images: [],
    };

    // set necessary bindings
    this.onFileUpload = this.onFileUpload.bind(this);
  }

  componentDidMount() {
    this.props.loadImages();
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      images: nextProps.images,
    });
  }

  onFileUpload(e) {
    const files = e.target.files;

    if (files.length < 1) {
      this.props.storeImage(null);
    } else {
      this.props.storeImage(files[0]);
    }
  }

  render() {
    const FileInput = (props) => (
      <input
        type="file"
        onChange={this.onFileUpload}
        className={props.className}
      />
    );

    const StyledFileInput = styled(FileInput)`
      opacity: 0;
      z-index: 1;
      width: 100px;
      margin-left: 10px;
      display: inline-block;
      position: absolute;
      height: 120px;
    `;

    const StyledImageButton = styled.div`
      background: no-repeat center url('https://image.flaticon.com/icons/svg/61/61112.svg');
      background-size: 100%;
      width: 100px;
      height: 100px;
      margin: 10px;
      z-index: 2;
    `;

    const { images, loadImagesError, storeImageError } = this.props;

    return (
      <div>
        {loadImagesError && <Label>
          <FormattedMessage {...messages.loadImagesError} />
        </Label>}
        {storeImageError && <Label>
          <FormattedMessage {...messages.storeImageError} />
        </Label>}
        <Row>
          {images.map((image, index) =>
            (<Column large={3} key={index}>
              <Image
                source={image}
              />
            </Column>)
          )}
          <Column large={3}>
            <StyledFileInput />
            <StyledImageButton />
          </Column>
        </Row>
      </div>
    );
  }
}

ImageList.propTypes = {
  loadImages: PropTypes.func.isRequired,
  storeImage: PropTypes.func.isRequired,
  images: PropTypes.any.isRequired,
  loadImagesError: PropTypes.bool.isRequired,
  storeImageError: PropTypes.bool.isRequired,
};

export default ImageList;
