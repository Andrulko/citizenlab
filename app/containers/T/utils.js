export function findTranslatedText(value, userLocale, tenantLocales) {
  let text = '';
  if (value[userLocale]) {
    text = value[userLocale];
  } else {
    tenantLocales.some((tenantLocale) => {
      if (value[tenantLocale]) {
        text = value[tenantLocale];
        return true;
      }
      return false;
    });
  }

  return text;
}
