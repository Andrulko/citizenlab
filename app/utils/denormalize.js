
// denormalize takes a resource out of the store and adds the data of all its relationships to it, if that data is available in the store
// Only use this function in your selectors if you're planning on actually using most of the relationships. If not, it's probably more efficient to get the relationship data at the moment you need it.

export default function denormalize(resources, type, id) {
  let resource = resources.getIn([type, id]);
  resource.get('relationships').forEach((ships, relation) => {
    const data = ships.get('data');
    if (data.constructor.name === 'List') {
      data.forEach((ship, i) => {
        const shipId = ship.get('id');
        const shipType = ship.get('type');
        const linkedResource = resources.getIn([shipType, shipId]);
        resource = resource.updateIn(['relationships', relation, 'data'], (innerData) => (
          innerData.set(i, linkedResource)
        ));
      });
    } else if (data.constructor.name === 'Map') {
      const ship = data;
      const shipId = ship.get('id');
      const shipType = ship.get('type');
      const linkedResource = resources.getIn([shipType, shipId]);
      resource = resource.setIn(['relationships', relation], linkedResource);
    }
  });
  return resource;
}
