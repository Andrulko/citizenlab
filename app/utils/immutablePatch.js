import { Map } from 'immutable'

Map.prototype.reactMap = function(func) {
  this.reduce((out, ...reduceArgs) => (
    out.concat(func(reduceArgs))
  ), []);
};
