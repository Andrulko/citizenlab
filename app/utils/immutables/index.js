export const getFromState = (state, property, subProperty) => {
  if (subProperty) {
    return state.getIn([property, subProperty]);
  }
  return state.get(property);
};
