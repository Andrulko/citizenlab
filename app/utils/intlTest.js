import React from 'react';
import { IntlProvider, intlShape } from 'react-intl';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';

const messages = require('./../translations/en'); // en.json
const intlProvider = new IntlProvider({ locale: 'en', messages }, {});
const { intl } = intlProvider.getChildContext();

/**
 * When using React-Intl `injectIntl` on components, props.intl is required.
 */
function nodeWithIntlProp(node) {
  return React.cloneElement(node, { intl });
}

function shallowWithIntl(node) {
  return shallow(nodeWithIntlProp(node), { context: { intl } });
}

function createComponentWithIntl(children, props = { locale: 'en' }) {
  return renderer.create(
    <IntlProvider {...props}>
      {children}
    </IntlProvider>
  );
}

function mountWithIntl(node) {
  return mount(nodeWithIntlProp(node), {
    context: { intl },
    childContextTypes: { intl: intlShape },
  });
}

export {
  shallowWithIntl,
  mountWithIntl,
  createComponentWithIntl,
};
