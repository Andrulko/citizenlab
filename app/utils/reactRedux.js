import { connect } from 'react-redux';

export function preprocess(...args) {
  return connect(...args, { areMergedPropsEqual: () => true });
}
