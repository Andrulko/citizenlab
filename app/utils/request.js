import 'whatwg-fetch';
import * as withQuery from 'with-query';

/**
 * Parses the JSON returned by a network request
 *
 * @param  {object} response A response from a network request
 *
 * @return {object}          The parsed JSON from the request
 */
// function parseJSON(response) {
//   return response.json();
// }

/**
 * Checks if a network request came back fine, and throws an error if not
 *
 * @param  {object} response   A response from a network request
 *
 * @return {object|undefined} Returns either the response, or throws an error
 */
// function checkStatus(response) {
//   if (response.status >= 200 && response.status < 300) {
//     return response;
//   }
//
//   const error = new Error(response.statusText);
//   error.response = response;
//   throw error;
// }

export function getJwt() {
  try {
    return window.localStorage.getItem('jwt');
  } catch (err) {
    console.log("[DEBUG] err =", err); // eslint-disable-line
    return null;
  }
}

export function setJwt(jwt) {
  window.localStorage.setItem('jwt', jwt);
}

/**
 * Requests a URL, returning a promise
 *
 * @param  {string} url       The URL we want to request
 * @param  {object} [options] The options we want to pass to "fetch"
 *
 * @return {object}           The response data
 */
export default function request(url, data, options, queryParameters) {
  const jwt = getJwt();
  const defaultOptions = {
    headers: {
      'Content-Type': 'application/json',
    },
  };

  if (jwt) {
    defaultOptions.headers['Authorization'] = `Bearer ${jwt}`; // eslint-disable-line
  }

  if (data) {
    defaultOptions.body = JSON.stringify(data);
  }

  const urlWithParams = (queryParameters
    ? withQuery(url, queryParameters)
    : url);

  return fetch(urlWithParams, Object.assign(defaultOptions, options))
    .then((response) => Promise.all([response, response.json()]))
    .then((result) => {
      const response = result[0];
      const json = result[1];

      if (response.ok) {
        return json;
      }

      const error = new Error(response.statusText);
      error.json = json;
      throw error;
    });
}
