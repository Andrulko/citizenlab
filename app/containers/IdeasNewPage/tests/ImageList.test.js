import React from 'react';
import { mount, shallow } from 'enzyme';
import { matcher, serializer } from 'jest-styled-components';

import { mountWithIntl } from '../../../utils/intlTest';
import ImageList, { Images, StyledFileInput } from '../ImageList';

expect.extend(matcher);
expect.addSnapshotSerializer(serializer);

describe('<ImageList />', () => {
  const images = ['1', '2'];
  const jestFn = jest.fn();
  it('should receive images prop', () => {
    const wrapper = mountWithIntl(
      <ImageList
        loadImages={jestFn}
        storeImage={jestFn}
        images={images}
        storeImageError={false}
        loadImagesError={false}
      />
    );

    expect(wrapper.prop('images')).toEqual(images);
  });

  it('should render the right number of images', () => {
    const wrapper = mount(
      <Images images={images} />
    );
    expect(wrapper.find('Image')).toHaveLength(2);
  });

  it('input button should render with the correct styles', () => {
    const wrapper = shallow(
      <StyledFileInput onFileUpload={jestFn} />
    );
    expect(wrapper).toMatchStyledComponentsSnapshot();
  });
});
