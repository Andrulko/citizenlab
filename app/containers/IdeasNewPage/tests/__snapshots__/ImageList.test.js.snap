// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageList /> input button should render with the correct styles 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <FileInput
    className="jkJdwS"
    onFileUpload={[Function]}
/>,
  "nodes": Array [
    <FileInput
      className="jkJdwS"
      onFileUpload={[Function]}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance":  {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Styled(undefined)
        onFileUpload={[Function]}
/>,
      "_debugID": 39,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StyledComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {
          "__styled-components__": undefined,
        },
        "props": Object {
          "onFileUpload": [Function],
        },
        "refs": Object {},
        "state": Object {
          "generatedClassName": "jkJdwS",
          "theme": Object {},
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 26,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <FileInput
          className="jkJdwS"
          onFileUpload={[Function]}
/>,
        "_debugID": 40,
        "_renderedOutput": <FileInput
          className="jkJdwS"
          onFileUpload={[Function]}
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Styled(undefined)
    onFileUpload={[Function]}
/>,
}
`;
